crates/protocol/src/types/varint.rs

```diff
@@ -0,0 +1,143 @@
+//! The variable-length integer type. Backed by an [`i32`]
+
+use bytes::{Buf, BufMut};
+use tokio::io::{self, AsyncReadExt, AsyncWriteExt};
+
+use crate::{error, ProtocolType, ProtocolTypeRaw};
+
+pub const MAX_SIZE: usize = 5;
+
+#[derive(Debug, thiserror::Error, from_never::FromNever)]
+pub enum DecodeError {
+    #[error(transparent)]
+    Eof(#[from] error::Eof),
+    #[error("The var int is larger than {MAX_SIZE}")]
+    TooLarge,
+}
+
+#[derive(Debug, Clone)]
+pub struct VarInt(pub i32);
+
+impl VarInt {
+    pub const fn written_size(self) -> usize {
+        match self.0 {
+            0 => 1,
+            n => (31 - n.leading_zeros() as usize) / 7 + 1,
+        }
+    }
+}
+
+impl ProtocolType for VarInt {
+    type DecodeError = DecodeError;
+    type EncodeError = error::Infallible;
+
+    fn decode(buffer: &mut impl Buf) -> Result<Self, Self::DecodeError> {
+        let mut val = 0;
+
+        for i in 0..MAX_SIZE {
+            if buffer.remaining() == 0 {
+                Err(error::Eof)?;
+            }
+
+            let byte = buffer.get_u8();
+            val |= (byte as i32 & 0b01111111) << (i * 7);
+
+            if byte & 0b10000000 == 0 {
+                return Ok(Self(val));
+            }
+        }
+
+        Err(DecodeError::TooLarge)
+    }
+
+    fn encode(&self, buffer: &mut impl BufMut) -> Result<(), Self::EncodeError> {
+        let x = self.0 as u64;
+        let stage1 = (x & 0x000000000000007f)
+            | ((x & 0x0000000000003f80) << 1)
+            | ((x & 0x00000000001fc000) << 2)
+            | ((x & 0x000000000fe00000) << 3)
+            | ((x & 0x00000000f0000000) << 4);
+
+        let leading = stage1.leading_zeros();
+
+        let unused_bytes = (leading - 1) >> 3;
+        let bytes_needed = 8 - unused_bytes;
+
+        // set all but the last MSBs
+        let msbs = 0x8080808080808080;
+        let msbmask = 0xffffffffffffffff >> (((8 - bytes_needed + 1) << 3) - 1);
+
+        let merged = stage1 | (msbs & msbmask);
+        let bytes = merged.to_le_bytes();
+
+        buffer.put(unsafe { bytes.get_unchecked(..bytes_needed as usize) });
```

Comment:
Lets use the safe alternative for now and tune in in the future on demand.

Reply:
Can someone else rewrite this algorithm for me

Reply:
As for now, just replace `get_unchecked` with slice indexing (or call to `get`). We will get back to it in the future.

Reply:
Done

Reply:
Good point. Yet, I will try to find a better way in the future.